src/ZkFold/Base/Protocol/Protostar/Oracle.hs

```diff
@@ -1,6 +1,72 @@
+{-# LANGUAGE AllowAmbiguousTypes  #-}
+{-# LANGUAGE NoOverloadedStrings  #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+
 module ZkFold.Base.Protocol.Protostar.Oracle where
 
+import           Data.Char                                      (ord)
+import           Data.Map.Strict                                (Map)
+import qualified Data.Map.Strict                                as M
+import           Data.Proxy                                     (Proxy (..))
+import           GHC.Generics
+import           GHC.TypeLits
+import           Prelude                                        (($), (.), (<$>))
+import qualified Prelude                                        as P
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Symbolic.Algorithms.Hash.MiMC           (mimcHash2)
+import           ZkFold.Symbolic.Algorithms.Hash.MiMC.Constants (mimcConstants)
+
 class RandomOracle a b where
     oracle :: a -> b
+    default oracle :: (Generic a, RandomOracle' (Rep a) b) => a -> b
+    oracle = oracle' . from
+
+instance (Ring a, FromConstant P.Integer a) => RandomOracle P.Integer a where
+    oracle = oracle @a . fromConstant
+
+instance (AdditiveMonoid b, RandomOracle a b) => RandomOracle (Map k a) b where
+    oracle = oracle . M.elems
+
+instance Ring a => RandomOracle a a where
+    oracle a = mimcHash2 mimcConstants a zero zero
+
+instance (AdditiveMonoid b, RandomOracle a b) => RandomOracle [a] b where
+    oracle as = sum $ oracle <$> as
+
+instance {-# OVERLAPPABLE #-} (Generic a, RandomOracle' (Rep a) b) => RandomOracle a b where
+
+class RandomOracle' f b where
+    oracle' :: f a -> b
+
+instance RandomOracle' V1 b where
+    oracle' = P.undefined -- No oracle for uninhabited types
```

Comment:
Can this instance be removed, then? This will lead to better compilation errors, too